dvc/data/stage.py

```diff
@@ -247,27 +250,88 @@ def _get_staging(odb: "ObjectDB") -> "ObjectDB":
     return ReferenceObjectDB(fs, fs_path, state=state)
 
 
-def _load_from_state(odb, staging, fs_path, fs, name):
+def _stage_raw_dir_hash(
+    staging: "ReferenceObjectDB",
+    tree: "Tree",
+    fs_path: "AnyFSPath",
+    fs: "FileSystem",
+    hash_info: "HashInfo",
+) -> "Tree":
+    assert tree.fs
+    staging.add(tree.fs_path, tree.fs, hash_info)
+    for key, _, oid in tree:
+        staging.add(fs.path.join(fs_path, *key), tree.fs, oid, verify=False)
+
+    raw = staging.get(hash_info)
+    tree.fs_path = raw.fs_path
+    tree.fs = raw.fs
+    tree.hash_info = hash_info
+    return tree
+
+
+def _try_load(
+    odb: "ObjectDB",
+    hash_info: "HashInfo",
+    meta: "Meta",
+    name: str,
+    raw_dir_hash: bool = False,
+) -> Tuple["Meta", "HashFile"]:
     from dvc.objects.errors import ObjectFormatError
 
-    from . import check, load
+    from . import check as ocheck
+    from . import load
     from .tree import Tree
 
+    try:
+        if raw_dir_hash:
+            obj = Tree.load(odb, hash_info)
+        else:
+            obj = load(odb, hash_info)
+            ocheck(odb, obj, check_hash=False)
+
+        if isinstance(obj, Tree):
+            meta.nfiles = len(obj)
+        assert obj.hash_info.name == name
+    except (ObjectFormatError, FileNotFoundError):
+        obj = None
+
+    return meta, obj
+
+
+def _load_from_state(
+    odb: "ObjectDB",
+    staging: "ReferenceObjectDB",
+    fs_path: "AnyFSPath",
+    fs: "FileSystem",
+    name: str,
+) -> Tuple["ObjectDB", "Meta", "HashFile"]:
     state = odb.state
     meta, hash_info = state.get(fs_path, fs)
-    if hash_info:
-        for odb_ in (odb, staging):
-            if odb_.exists(hash_info):
-                try:
-                    obj = load(odb_, hash_info)
-                    check(odb_, obj, check_hash=False)
-                    if isinstance(obj, Tree):
-                        meta.nfiles = len(obj)
-                    assert obj.hash_info.name == name
-                    return odb_, meta, obj
-                except (ObjectFormatError, FileNotFoundError):
-                    pass
-    raise FileNotFoundError
+    if not hash_info:
+        raise FileNotFoundError
+
+    for odb_ in (odb, staging):
+        if not odb_.exists(hash_info):
+            continue
+
+        meta, obj = _try_load(odb_, hash_info, meta, name)
```

Comment:
You are passing the name and the meta just to modify and assert them there, while you could just
```suggestion
        meta, obj = _try_load(odb_, hash_info)
        if isinsatnce(obj, Tree):
            meta.nfiles = len(obj)
        assert obj.hash_info.name == name
```
and avoid redundant passthrough